File: src/lib/sources/duckduckgo.ts
import * as cheerio from 'cheerio';
import { RawSearchResult } from '@/types';
import { searchLimiter } from '@/lib/utils/rate-limiter';

export async function searchDuckDuckGo(keyword: string): Promise<RawSearchResult[]> {
  await searchLimiter.waitForSlot();

  const query = encodeURIComponent(keyword);
  const url = `https://html.duckduckgo.com/html/?q=${query}`;

  try {
    const res = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        'Accept': 'text/html',
      },
      signal: AbortSignal.timeout(8000),
    });

    if (!res.ok) return [];

    const html = await res.text();
    const $ = cheerio.load(html);
    const results: RawSearchResult[] = [];

    $('.result').each((_, el) => {
      const titleEl = $(el).find('.result__title a');
      const title = titleEl.text().trim();
      const href = titleEl.attr('href') || '';
      const snippet = $(el).find('.result__snippet').text().trim();

      if (title && href) {
        results.push({
          title,
          url: href.startsWith('//') ? `https:${href}` : href,
          snippet,
          source_type: 'duckduckgo',
        });
      }
    });

    return results.slice(0, 8);
  } catch (e) {
    console.error('[duckduckgo] Search failed:', e);
    return [];
  }
}

File: agent-skills/hot-monitor/scripts/sources/duckduckgo.py
"""Ported from src/lib/sources/duckduckgo.ts"""

from __future__ import annotations

import httpx
from bs4 import BeautifulSoup
from urllib.parse import quote

from rate_limiter import search_limiter
from models import RawSearchResult
from sources.bing import UA


async def search_duckduckgo(keyword: str) -> list[RawSearchResult]:
    await search_limiter.wait_for_slot()

    search_query = f'"{keyword}"' if " " in keyword else keyword
    url = f"https://html.duckduckgo.com/html/?q={quote(search_query)}&df=w"

    try:
        async with httpx.AsyncClient(timeout=8.0) as client:
            res = await client.get(
                url,
                headers={"User-Agent": UA, "Accept": "text/html"},
            )
        if res.status_code != 200:
            return []

        soup = BeautifulSoup(res.text, "html.parser")
        results: list[RawSearchResult] = []

        for el in soup.select(".result"):
            title_el = el.select_one(".result__title a")
            if not title_el:
                continue
            title = title_el.get_text(strip=True)
            href = title_el.get("href") or ""
            snippet_el = el.select_one(".result__snippet")
            snippet = snippet_el.get_text(strip=True) if snippet_el else ""
            if title and href:
                if href.startswith("//"):
                    href = f"https:{href}"
                results.append(
                    {
                        "title": title,
                        "url": href,
                        "snippet": snippet,
                        "source_type": "duckduckgo",
                    }
                )

        return results[:8]
    except Exception as e:
        print(f"[duckduckgo] Search failed: {e}", flush=True)
        return []
